### PR TITLE
Fix kwargs passing on Ruby 3.0 for sucker_punch

### LINKED ISSUE #1372: sucker_punch instrumentation broken for keyword args on Ruby 3
While investigating #1369 I half-suspected that once `sucker_punch` was fixed upstream that our instrumentation would not be correct, and now that `sucker_punch` 3.0.1 is out and does it correctly, I coded up a quick test that shows this:

```ruby
puts RUBY_DESCRIPTION

require 'sucker_punch'

class FooJob
  include SuckerPunch::Job

  def perform(*args, required:)
    puts [args, required].inspect
  end
end

FooJob.perform_async(1, required: 2)
FooJob.perform_in(0, 1, required: 2)
FooJob.__run_perform(1, required: 2)

sleep 1

require 'ddtrace'

Datadog.configure do |c|
  c.use :sucker_punch
end

FooJob.perform_async(1, required: 2)
FooJob.perform_in(0, 1, required: 2)
FooJob.__run_perform(1, required: 2)

sleep 1
```

Behavior on Ruby 2.x:

```ruby
$ bundle exec ruby sucker_punch_instrumentation.rb
ruby 2.7.2p137 (2020-10-01 revision 5445e04352) [x86_64-darwin19]
[[1], 2]
[[1], 2]
[[1], 2]
[[1], 2]
[[1], 2]
[[1], 2]
```

Behavior on Ruby 3.x:

```ruby
$ bundle exec ruby sucker_punch_instrumentation.rb
ruby 3.0.0p0 (2020-12-25 revision 95aff21468) [x86_64-darwin19]
[[1], 2]
[[1], 2]
[[1], 2]
E, [2021-02-18T09:00:41.528096 #15995] ERROR -- : Sucker Punch job error for class: 'FooJob' args: [1, {:required=>2}]
ArgumentError missing keyword: :required
sucker_punch_instrumentation.rb:8:in `perform'
/Users/ivo.anjo/.rvm/gems/ruby-3.0.0/gems/sucker_punch-3.0.1/lib/sucker_punch/job.rb:63:in `__run_perform'
/Users/ivo.anjo/datadog/dd-trace-rb/lib/ddtrace/contrib/sucker_punch/instrumentation.rb:32:in `block in __run_perform'
/Users/ivo.anjo/datadog/dd-trace-rb/lib/ddtrace/contrib/sucker_punch/instrumentation.rb:77:in `block in __with_instrumentation'
/Users/ivo.anjo/datadog/dd-trace-rb/lib/ddtrace/tracer.rb:284:in `trace'
/Users/ivo.anjo/datadog/dd-trace-rb/lib/ddtrace/contrib/sucker_punch/instrumentation.rb:73:in `__with_instrumentation'
/Users/ivo.anjo/datadog/dd-trace-rb/lib/ddtrace/contrib/sucker_punch/instrumentation.rb:21:in `__run_perform'
sucker_punch_instrumentation.rb:27:in `<main>'
E, [2021-02-18T09:00:41.528391 #15995] ERROR -- : Sucker Punch job error for class: 'FooJob' args: [1, {:required=>2}]
ArgumentError missing keyword: :required
sucker_punch_instrumentation.rb:8:in `perform'
/Users/ivo.anjo/.rvm/gems/ruby-3.0.0/gems/sucker_punch-3.0.1/lib/sucker_punch/job.rb:63:in `__run_perform'
/Users/ivo.anjo/datadog/dd-trace-rb/lib/ddtrace/contrib/sucker_punch/instrumentation.rb:32:in `block in __run_perform'
/Users/ivo.anjo/datadog/dd-trace-rb/lib/ddtrace/contrib/sucker_punch/instrumentation.rb:77:in `block in __with_instrumentation'
/Users/ivo.anjo/datadog/dd-trace-rb/lib/ddtrace/tracer.rb:284:in `trace'
/Users/ivo.anjo/datadog/dd-trace-rb/lib/ddtrace/contrib/sucker_punch/instrumentation.rb:73:in `__with_instrumentation'
/Users/ivo.anjo/datadog/dd-trace-rb/lib/ddtrace/contrib/sucker_punch/instrumentation.rb:21:in `__run_perform'
/Users/ivo.anjo/.rvm/gems/ruby-3.0.0/gems/sucker_punch-3.0.1/lib/sucker_punch/job.rb:38:in `block in perform_async'
/Users/ivo.anjo/.rvm/gems/ruby-3.0.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:363:in `run_task'
/Users/ivo.anjo/.rvm/gems/ruby-3.0.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:352:in `block (3 levels) in create_worker'
/Users/ivo.anjo/.rvm/gems/ruby-3.0.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:335:in `loop'
/Users/ivo.anjo/.rvm/gems/ruby-3.0.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:335:in `block (2 levels) in create_worker'
/Users/ivo.anjo/.rvm/gems/ruby-3.0.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:334:in `catch'
/Users/ivo.anjo/.rvm/gems/ruby-3.0.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:334:in `block in create_worker'
E, [2021-02-18T09:00:41.529427 #15995] ERROR -- : Sucker Punch job error for class: 'FooJob' args: [1, {:required=>2}]
ArgumentError missing keyword: :required
sucker_punch_instrumentation.rb:8:in `perform'
/Users/ivo.anjo/.rvm/gems/ruby-3.0.0/gems/sucker_punch-3.0.1/lib/sucker_punch/job.rb:63:in `__run_perform'
/Users/ivo.anjo/datadog/dd-trace-rb/lib/ddtrace/contrib/sucker_punch/instrumentation.rb:32:in `block in __run_perform'
/Users/ivo.anjo/datadog/dd-trace-rb/lib/ddtrace/contrib/sucker_punch/instrumentation.rb:77:in `block in __with_instrumentation'
/Users/ivo.anjo/datadog/dd-trace-rb/lib/ddtrace/tracer.rb:284:in `trace'
/Users/ivo.anjo/datadog/dd-trace-rb/lib/ddtrace/contrib/sucker_punch/instrumentation.rb:73:in `__with_instrumentation'
/Users/ivo.anjo/datadog/dd-trace-rb/lib/ddtrace/contrib/sucker_punch/instrumentation.rb:21:in `__run_perform'
/Users/ivo.anjo/.rvm/gems/ruby-3.0.0/gems/sucker_punch-3.0.1/lib/sucker_punch/job.rb:46:in `block in perform_in'
/Users/ivo.anjo/.rvm/gems/ruby-3.0.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/executor/safe_task_executor.rb:24:in `block in execute'
/Users/ivo.anjo/.rvm/gems/ruby-3.0.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/synchronization/mutex_lockable_object.rb:41:in `block in synchronize'
/Users/ivo.anjo/.rvm/gems/ruby-3.0.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/synchronization/mutex_lockable_object.rb:41:in `synchronize'
/Users/ivo.anjo/.rvm/gems/ruby-3.0.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/synchronization/mutex_lockable_object.rb:41:in `synchronize'
/Users/ivo.anjo/.rvm/gems/ruby-3.0.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/executor/safe_task_executor.rb:19:in `execute'
/Users/ivo.anjo/.rvm/gems/ruby-3.0.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/ivar.rb:169:in `safe_execute'
/Users/ivo.anjo/.rvm/gems/ruby-3.0.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/scheduled_task.rb:285:in `process_task'
/Users/ivo.anjo/.rvm/gems/ruby-3.0.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/executor/timer_set.rb:98:in `block in ns_post_task'
/Users/ivo.anjo/.rvm/gems/ruby-3.0.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:363:in `run_task'
/Users/ivo.anjo/.rvm/gems/ruby-3.0.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:352:in `block (3 levels) in create_worker'
/Users/ivo.anjo/.rvm/gems/ruby-3.0.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:335:in `loop'
/Users/ivo.anjo/.rvm/gems/ruby-3.0.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:335:in `block (2 levels) in create_worker'
/Users/ivo.anjo/.rvm/gems/ruby-3.0.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:334:in `catch'
/Users/ivo.anjo/.rvm/gems/ruby-3.0.0/gems/concurrent-ruby-1.1.8/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:334:in `block in create_worker'
```

The fix is probably quite simple: sprinkle `ruby2_keywords :foo if respond_to?(:ruby2_keywords, true)` on a bunch of our methods, see https://eregon.me/blog/2021/02/13/correct-delegation-in-ruby-2-27-3.html , and add tests to cover it ;)

### PR DESCRIPTION
**Pull request by lloeki**
The kwargs separation change for Ruby 3.0 is tricky: there are different behaviours on 2.6, 2.7, and 3.0. While `ruby2_keywords` attempts to smooth that out, it is due to disappear on 3.2, therefore relying on it produces surefire broken code when 3.2 is released and 2.6 EOL'd.

`ruby2_keywords` also suffers from subtle brokenness by virtue of leveraging an internal stateful flag attached to the hash. This is generally OK for very local uses but gets troublesome with complex delegations and even more for instrumentors. Case in point, attempting to use `ruby2_keywords` here works for `perform_in` and `perform_async`, but does not solve the issue in the `__run_perform` direct call case.

Thus the only future-proof and reliable syntax on Ruby >= 3 is to capture the arguments with `*args, **kwargs` and forward them the same way, without `ruby2_keywords`.

Conversely, the only working syntax on Ruby < 3 is capturing both with `*args` and forwarding them the same way, using `ruby2_keywords` on 2.7 to restore the previous behaviour.

Therefore dynamic function definition per version is sadly necessary because of the syntactic difference in argument passing and forwarding.

Fixes #1372

### CODE REVIEW

**Comment by ivoanjo** [comment]
> While ruby2_keywords attempts to smooth that out, it is due to disappear on 3.2

Really? :o

I hadn't spotted this, do you remember where you saw it originally? I'd be really curious to follow the discussion.

**Comment by lloeki** [comment]
> Really? :o

Yep indeed, straight from [the source](https://www.ruby-lang.org/en/news/2019/12/12/separation-of-positional-and-keyword-arguments-in-ruby-3-0/) :)

> If you really worry about the portability, use ruby2_keywords. (Acknowledge that Ruby 2.6 or before themselves have tons of corner cases in keyword arguments. :-) ruby2_keywords might be removed in the future after Ruby 2.6 reaches end-of-life. At that point, we recommend to explicitly delegate keyword arguments (see Ruby 3 code above).

Sadly I can't remember where I saw the "might" becoming less evasive and a more definitive "shall", including mention of Ruby 3.2 (the one which will EOL 2.6).

**Comment by lloeki** [comment]
Silly me made a mistake on `perform_async`, so `__run_perform` was in fact unaffected (thanks @ivoanjo!).

I dropped the complexity down by removing the `*args, **kwargs` cases but introduced a version cap to limit max version to 3.2, guaranteed to have `ruby2_keywords`, thus not break upon upstream release. This should also cover other `ruby2_keyword` usage in the codebase.

I'm interested in feedback about which would be preferable (version cap or kwargs syntax behind a conditional).

File: lib/ddtrace/version.rb
Comment (marcotc) [inline lib/ddtrace/version.rb:14]: I haven't been able to find information about the removal of `ruby2_keywords` in Ruby 3.2, could you point me to any sources on it?

**Review by ivoanjo — CHANGES_REQUESTED** [review]
Looks good to me! I think the extra context is really valuable to make it extremely clear why the maximum version is there, but otherwise this is good to go :)

File: lib/ddtrace/version.rb
Comment (lloeki) [reply lib/ddtrace/version.rb:14]: See this comment: https://github.com/DataDog/dd-trace-rb/pull/1495#issuecomment-833435661

I also [asked Benoit Daloze](https://twitter.com/eregontp/status/1390390407457034243), TruffleRuby lead and one of the regularly involved persons in the kwargs subject on [bugs.ruby-lang.org](https://bugs.ruby-lang.org/users/772):

>> Given that `respond_to?(:ruby2_keywords)` would become false around Ruby 2.6 EOL, wouldn't the correct AND future-proof way be to `if RUBY_VERSION < '3'` that `*args` section, and use `*args, **kwargs` on the `else` clause?
>
> Yes, that's technically correct, so this would work:
>
> ```
> if RUBY_VERSION < "3.0"
>   def foo(*args, &block)
>     target(*args, &block)
>   end
>   ruby2_keywords :foo if respond_to?(:ruby2_keywords, true)
> else
>   def foo(*args, **kwargs, &block)
>     target(*args, **kwargs, &block)
>   end
> end
> ```
>
> However, the idea of ruby-core is that `ruby2_keywords` would only be removed when 2.6 is EOL and unsupported by most gems, or in other words when gems can assume Ruby >= 3.0, and then only the second part is needed. That strategy is really about avoiding the duplicated code.
>
> So ruby-core's view is change (*args) to (*args)+ruby2_kewords so it works on Ruby 2.7+, and later on (when your app/gem can assume Ruby 3+) change (*args)+ruby2_kewords to (*args, **kwargs).

Not sure if Benoit is core but he's at least core contributor and very much in touch. Anyway, I figured out, I could just as well [ask Matz](https://twitter.com/lloeki/status/1390651699958882304) too for a clarification on the timetable.

File: lib/ddtrace/version.rb
Comment (lloeki) [reply lib/ddtrace/version.rb:16]: `< 3.2` would match `3.2.0-preview1`, mistakenly so (we want it excluded). Thats how `rubygems` [encodes the minimal version including `prerelease?` ones](https://github.com/rubygems/rubygems/blob/c5c4c62547444330c170ab227a8798bc503c6db5/lib/rubygems/version.rb#L26-L29) (interestingly enough, ["no version requirement" is in fact `Gem::Version.new('0.a')`](https://github.com/rubygems/rubygems/blob/c5c4c62547444330c170ab227a8798bc503c6db5/lib/rubygems/requirement.rb#L40)).

Trivia: interestingly enough, because of the `-p648`:

```
irb> Gem::Version.new('2.0.0p648').prerelease?
=> true
irb> Gem::Version.new('2.0.0p648') < Gem::Version.new('2.0')
=> true
```

Which, as you see, had the potency to cause shenanigans in comparisons if you weren't careful ;)

Comment (ivoanjo) [reply lib/ddtrace/version.rb:16]: Thanks! I totally did now know that.

TBH, I don't quite mind if people use it with preview versions -- it's a Ruby preview version, so you know going in that "things can break" -- so I'd still go with the 3.2, but I think 3.2.a is very reasonable as well so I'll let you make the final decision :)

Comment (lloeki) [reply lib/ddtrace/version.rb]: Ultimately I went for `3.2`, as I thought they might eventually delay the drop if enough gems have not made the move. This way this can be easily tested for that before the final 3.2 release, as well as not blocking early testing of non-kwargs dependent integrations on the preview.

**Review by ivoanjo — APPROVED** [review]
Awesome 🚀 💯 👍

**Pull request merged**
